rcgen/src/lib.rs

```diff
@@ -1593,6 +1585,24 @@ impl Certificate {
 	pub fn serialize_private_key_pem(&self) -> String {
 		self.key_pair.serialize_pem()
 	}
+
+	/// Return a subject [`KeyPair`] to use for the [`CertificateParams`].
+	///
+	/// If `params` do not specify a [`KeyPair`], a new one is generated.
+	/// If `params` specifies a [`KeyPair`], that isn't compatible with the signature
+	/// algorithm specified, an error is returned.
+	fn prepare_key_pair(
```

Comment:
Maybe this could be `KeyPair::prepare()`?

Reply:
We don't have a `KeyPair` in this case, we have an `Option<KeyPair>` that might be `None`. Most of this helper exists to deal with the `None` case, there doesn't seem to be much of value to lift into a `KeyPair` fn for the non-none case so I left this as is.

Reply:
We're not using any data from Certificate so it seems weird for this to be part of the Certificate impl. Plus it yields a Result<KeyPair, Error> so it does feel like a kind of constructor for KeyPair except that you might inject a pre-existing one to verify its algorithm. Maybe call it KeyPair::validate_or_generate()?

Reply:
You've won me over :+1: I updated the implementation to associate this with the `KeyPair` type as `KeyPair::validate_or_generate`.